app/src/main/java/com/lightricks/feedexercise/data/FeedDataSource.kt

```diff
@@ -0,0 +1,52 @@
+package com.lightricks.feedexercise.data
+
+import com.lightricks.feedexercise.database.FeedDao
+import com.lightricks.feedexercise.domain.FeedError
+import com.lightricks.feedexercise.domain.FeedSuccess
+import com.lightricks.feedexercise.network.FeedApiService
+import com.lightricks.feedexercise.network.UrlProvider
+import com.lightricks.feedexercise.util.result.Result
+import com.lightricks.feedexercise.util.result.toFailureResult
+import com.lightricks.feedexercise.util.result.toSuccessResult
+import java.net.UnknownHostException
+
+interface FeedDataSource {
+    suspend fun getFeed(): Result<FeedError, FeedSuccess>
+}
+
+internal class RetrofitFeedDataSource(
+    private val apiService: FeedApiService,
+    private val thumbnailBaseUrlProvider: UrlProvider
+) : FeedDataSource {
+    override suspend fun getFeed(): Result<FeedError, FeedSuccess> {
+        return try {
+            val response = apiService.getFeed()
+            return when {
+                response.isSuccessful && response.body() != null -> requireNotNull(response.body())
+                    .toFeed(thumbnailBaseUrlProvider)
+                    .let { feed -> FeedSuccess(feed = feed, isCache = false) }
+                    .toSuccessResult()
+                else -> FeedError.NetworkError.toFailureResult()
+            }
+        } catch (e: UnknownHostException) {
+            FeedError.NetworkError.toFailureResult()
+        } catch (e: Throwable) {
+            FeedError.UnknownError.toFailureResult()
+        }
+    }
+}
+
+internal class RoomFeedDataSource(
+    private val dao: FeedDao
+) : FeedDataSource {
+    override suspend fun getFeed(): Result<FeedError, FeedSuccess> {
```

Comment:
For the cache the api should return a flow since we want to keep updated each time the cache data change.

Reply:
Also, IMO this class should have the `save` method as well. Having the two functionalities separated make thing harder to follow, so having all the functionality of saving and retrieving the data in the same place is better IMO.